keps/sig-network/2594-multiple-cluster-cidrs/README.md

```diff
@@ -0,0 +1,992 @@
+        and store this association with the node.
+    -   If `PodCIDR` is set, but is not part of one of the tracked
+        `ClusterCIDRConfig`, emit a K8s event but do nothing.
+    -   If `PodCIDR` is not set, save Node for allocation in the next step.
+        After processing all nodes, allocate ranges to any nodes without Pod
+        CIDR(s) [Same logic as Node Added event]
+
+#### Reconciliation Loop
+
+This go-routine will watch for cleanup operations and failed allocations and
+continue to try them in the background.
+
+For example if a Node can't be allocated a PodCIDR, it will be periodically
+retried until it can be allocated a range or it is deleted.
+
+#### Event Watching Loops
+
+##### Node Added
+
+If the Node already has a `PodCIDR` allocated, mark the CIDRs as used.
+
+Otherwise, go through the list of `ClusterCIDRConfig`s and find ranges matching
+the node selector from each family. Attempt to allocate Pod CIDR(s) with the
+given per-node size. If that `ClusterCIDRConfig` cannot fit a node, search for
+another `ClusterCIDRConfig`.
+
+If no `ClusterCIDRConfig` matches the node, or if all matching
+`ClusterCIDRConfig`s are full, raise a K8s event and put the Node on the
+reconciliation queue (infinite retries). Upon successfully allocating CIDR(s),
+update the node object with the podCIDRs.
+
+##### Node Updated
+
+Check that its Pod CIDR(s) match internal allocation.
+
+-   If node.spec.PodCIDRs is already filled up, honor that allocation and mark
+    those ranges as allocated.
+-   If the node.spec.PodCIDRs is filled with a CIDR not from any
+    `ClusterCIDRConfig`, raise a K8sEvent.
+-   If the ranges are already marked as allocated for some other node, raise
+    another error event (there isn’t an obvious reconciliation step the
+    controller can take unilaterally).
+
+##### Node Deleted
+
+Release said Node’s allocation from the internal data-structure.
+
+If this Node is the last one using a particular `ClusterCIDRConfig` that has
+been slated for deletion, trigger the deletion flow again (so that the finalizer
+is removed and internal data structures are cleaned up).
+
+##### ClusterCIDRConfig Added
+
+Install a finalizer on the `ClusterCIDRConfig` called "networking.kubernetes.io/cluster-cidr-config-finalizer".
+
+Update internal representation of CIDRs to include the new range. Every failed
+Node Allocation is stored in a queue, that will be tried again with the new
+range by the reconciliation loop.
+
+##### ClusterCIDRConfig Updated
+
+_`CIDR`, and `PerNodeMaskSize` are immutable_
+
+Update any internal state as required.
+
+If the `NodeSelector` changed to include a node that is currently awaiting a
+PodCIDR allocation, it will be allocated by the reconiliation loop. 
+
+If the `NodeSelector` changes to unselect a Node currently using that range,
+there is no change to the Node's allocation. Users will have to manually
+restart/recreated that Node to reset it's `PodCIDR` allocation.
```

Comment:
is this worth the effort?
is not better to play with the labels on the nodes instead?

Reply:
That's a fair point. I was thinking we could use this for swaps, but since the finalizer will keep the old range around while it's in use, you can delete the old range, create a new one with the same selector and the appropriate CIDR, and then start restarting the nodes.

I guess the only benefit to making it editable is it lets users have more flexibility -- but maybe we should start with un-editable and then open it up if really needed.